tests/integration/test_applications_send_logs.py

```diff
@@ -0,0 +1,127 @@
+            resources=tester_resources,
+            application_name="log-proxy-tester-syslog",
+            config={"syslog": "true", "file_forwarding": "false"},
+        ),
+    )
+    await ops_test.model.wait_for_idle(apps=app_names, status="active")
+
+    for t in tester_app_names:
+        await ops_test.model.add_relation(loki_app_name, t)
+    await ops_test.model.wait_for_idle(apps=app_names, status="active")
+
+    # This is silly, but we need to actually wait to get log data
+    time.sleep(30)
+
+    for query in tester_apps.values():
+        logs = await loki_api_query(ops_test, loki_app_name, query)
+        assert len(logs["values"]) > 0
+
+
+@pytest.mark.abort_on_fail
+async def test_scale_up_also_gets_logs(ops_test):
+    await ops_test.model.applications[loki_app_name].scale(scale=3)
+    await ops_test.model.wait_for_idle(
+        apps=[loki_app_name], status="active", timeout=600, wait_for_exact_units=0
+    )
+    assert await is_loki_up(ops_test, loki_app_name, num_units=3)
+
+    # Trigger a log message to fire an alert on just to ensure we have logs
+    await ops_test.model.applications["loki-tester"].units[0].run_action(
+        "log-error", message="Error logged!"
+    )
+    await ops_test.model.wait_for_idle(
+        apps=app_names, status="active", timeout=1000, idle_period=60
+    )
+
+    assert await is_loki_up(ops_test, loki_app_name, num_units=3)
+
+    # This is silly, but we need to actually wait to get log data
+    time.sleep(30)
+
+    for query in tester_apps.values():
+        logs_per_unit = await asyncio.gather(
+            loki_api_query(ops_test, loki_app_name, query, unit_num=0),
+            loki_api_query(ops_test, loki_app_name, query, unit_num=1),
+            loki_api_query(ops_test, loki_app_name, query, unit_num=2),
+        )
+        assert all(len(logs["values"]) > 0 for logs in logs_per_unit)
+
+
+@pytest.mark.abort_on_fail
+async def test_logs_persist_after_upgrade(ops_test, loki_charm):
+    # Refresh from path
+    await ops_test.model.applications[loki_app_name].refresh(path=loki_charm, resources=resources)
+    await ops_test.model.wait_for_idle(
+        apps=app_names, status="active", timeout=1000, idle_period=60
+    )
+    assert await is_loki_up(ops_test, loki_app_name, num_units=3)
+
+    # This is silly, but we need to actually wait to get log data
+    time.sleep(30)
+
+    # Trigger a log message to fire an alert on just to ensure we have logs
+    await ops_test.model.applications["loki-tester"].units[0].run_action(
+        "log-error", message="Error logged!"
+    )
+
+    for query in tester_apps.values():
+        logs_per_unit = await asyncio.gather(
+            loki_api_query(ops_test, loki_app_name, query, unit_num=0),
+            loki_api_query(ops_test, loki_app_name, query, unit_num=1),
+            loki_api_query(ops_test, loki_app_name, query, unit_num=2),
+        )
+        assert all(len(logs["values"]) > 0 for logs in logs_per_unit)
```

Comment:
Seems like the `time.sleep(30)` should be removed, otherwise this will always pass.
I.e. to check persistence across upgrade, we'd need to assert against the log content before any additions are made.

Reply:
The point of the test here is "after upgrade, do we still get logs which match the same labels?" The sleep has nothing to do with that, really, except for the `endpoints` timing above.

I'll add a count which can be compared before and after though.